restapi/admin_heal.go

```diff
@@ -0,0 +1,378 @@
+	itemStatus.Before.Missing, itemStatus.After.Missing = i.GetMissingCounts()
+	itemStatus.Before.Corrupted, itemStatus.After.Corrupted = i.GetCorruptedCounts()
+	itemStatus.Before.Offline, itemStatus.After.Offline = i.GetOfflineCounts()
+	itemStatus.Before.Drives = i.Before.Drives
+	itemStatus.After.Drives = i.After.Drives
+	h.ItemsHealthStatus = append(h.ItemsHealthStatus, itemStatus)
+	h.HealthBeforeCols[beforeColor]++
+	h.HealthAfterCols[afterColor]++
+	return nil
+}
+
+// getObjectHCCChange - returns before and after color change for
+// objects
+func getObjectHCCChange(h madmin.HealResultItem) (b, a col, err error) {
+	parityShards := h.ParityBlocks
+	dataShards := h.DataBlocks
+
+	onlineBefore, onlineAfter := h.GetOnlineCounts()
+	surplusShardsBeforeHeal := onlineBefore - dataShards
+	surplusShardsAfterHeal := onlineAfter - dataShards
+
+	b, err = getHColCode(surplusShardsBeforeHeal, parityShards)
+	if err != nil {
+		return
+	}
+	a, err = getHColCode(surplusShardsAfterHeal, parityShards)
+	return
+
+}
+
+// getReplicatedFileHCCChange - fetches health color code for metadata
+// files that are replicated.
+func getReplicatedFileHCCChange(h madmin.HealResultItem) (b, a col, err error) {
+	getColCode := func(numAvail int) (c col, err error) {
+		// calculate color code for replicated object similar
+		// to erasure coded objects
+		quorum := h.DiskCount/h.SetCount/2 + 1
+		surplus := numAvail/h.SetCount - quorum
+		parity := h.DiskCount/h.SetCount - quorum
+		c, err = getHColCode(surplus, parity)
+		return
+	}
+
+	onlineBefore, onlineAfter := h.GetOnlineCounts()
+	b, err = getColCode(onlineBefore)
+	if err != nil {
+		return
+	}
+	a, err = getColCode(onlineAfter)
+	return
+}
+
+func getHColCode(surplusShards, parityShards int) (c col, err error) {
+	if parityShards < 1 || parityShards > 8 || surplusShards > parityShards {
+		return c, fmt.Errorf("invalid parity shard count/surplus shard count given")
+	}
+	if surplusShards < 0 {
+		return colGrey, err
+	}
+	colRow := hColTable[parityShards]
+	for index, val := range colRow {
+		if val != -1 && surplusShards <= val {
+			return hColOrder[index], err
+		}
+	}
+	return c, fmt.Errorf("cannot get a heal color code")
+}
+
+func getHRITypeAndName(i madmin.HealResultItem) (typ, name string) {
+	name = fmt.Sprintf("%s/%s", i.Bucket, i.Object)
+	switch i.Type {
+	case madmin.HealItemMetadata:
+		typ = "system"
+		name = i.Detail
+	case madmin.HealItemBucketMetadata:
+		typ = "system"
+		name = "bucket-metadata:" + name
+	case madmin.HealItemBucket:
+		typ = "bucket"
+	case madmin.HealItemObject:
+		typ = "object"
+	default:
+		typ = fmt.Sprintf("!! Unknown heal result record %#v !!", i)
+		name = typ
+	}
+	return
+}
+
+// getHealOptionsFromReq return options from request for healing process
+// path come as : `/heal/bucket1` and query params come on request form
+func getHealOptionsFromReq(req *http.Request) (hOptions healOptions, err error) {
```

Comment:
I think we should return pointer to struct to be consistent with the functions called by the other endpoints

```suggestion
func getHealOptionsFromReq(req *http.Request) (hOptions *healOptions, err error) {
```

Reply:
done